x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,282 @@
+	Payload            []byte `json:"payload"`
+	Type               int    `json:"type"`
+}
+
+func validateMessage(ctx sdk.Context, k keeper.Keeper, n types.Nexus, ibcPath string, msg Message, token sdk.Coin) error {
+	chainName, ok := k.GetChainNameByIBCPath(ctx, ibcPath)
+	if !ok {
+		return fmt.Errorf("unrecognized IBC path %s", ibcPath)
+	}
+	chain := funcs.MustOk(n.GetChain(ctx, chainName))
+
+	if !n.IsChainActivated(ctx, chain) {
+		return fmt.Errorf("chain %s registered for IBC path %s is deactivated", chain.Name, ibcPath)
+	}
+
+	destChainName := nexus.ChainName(msg.DestinationChain)
+	if err := destChainName.Validate(); err != nil {
+		return err
+	}
+
+	destChain, ok := n.GetChain(ctx, destChainName)
+	if !ok {
+		return fmt.Errorf("unrecognized destination chain %s", destChain.Name)
+	}
+
+	if !n.IsChainActivated(ctx, destChain) {
+		return fmt.Errorf("chain %s is deactivated", destChain.Name)
+	}
+
+	if err := n.ValidateAddress(ctx, nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}); err != nil {
+		return err
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		return nil
+	case nexus.TypeGeneralMessageWithToken, nexus.TypeSendToken:
+		if !n.IsAssetRegistered(ctx, chain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), destChain.Name)
+		}
+
+		if !n.IsAssetRegistered(ctx, destChain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), destChain.Name)
+		}
+		return nil
+	default:
+		return fmt.Errorf("unrecognized message type")
+	}
+}
+
+// OnRecvMessage handles general message from a cosmos chain
+func OnRecvMessage(ctx sdk.Context, k keeper.Keeper, ibcK keeper.IBCKeeper, n types.Nexus, b types.BankKeeper, packet ibcexported.PacketI) ibcexported.Acknowledgement {
+	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+	data, err := toICS20Packet(packet)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// skip if packet not sent to Axelar message sender account
+	if data.GetReceiver() != types.MessageSender.String() {
+		return ack
+	}
+
+	var msg Message
+	if err := json.Unmarshal([]byte(data.GetMemo()), &msg); err != nil {
+		ackErr := sdkerrors.Wrapf(types.ErrGeneralMessage, "cannot unmarshal memo")
+		return channeltypes.NewErrorAcknowledgement(ackErr)
+	}
+
+	// extract token from packet
+	token := extractTokenFromPacketData(packet)
+
+	path := types.NewIBCPath(packet.GetDestPort(), packet.GetDestChannel())
+
+	if err := validateMessage(ctx, k, n, path, msg, token); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	sourceChain := funcs.MustOk(k.GetChainNameByIBCPath(ctx, path))
+	sourceAddress := nexus.CrossChainAddress{
+		Chain:   funcs.MustOk(n.GetChain(ctx, sourceChain)),
+		Address: data.GetSender(),
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		err = handleMessage(ctx, n, sourceAddress, msg)
```

Comment:
we'll park the token in the gmp account if it's a call contract? Should we allow someone to specify a local receiver account in the msg?